src/charm.py

```diff
@@ -7,20 +7,52 @@
 import logging
 
 import ops
+from charms.operator_libs_linux.v2 import snap
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+
+UPF_SNAP_NAME = "sdcore-upf"
+UPF_SNAP_CHANNEL = "latest/edge"
+UPF_SNAP_REVISION = "3"
+
 
 logger = logging.getLogger(__name__)
 
 
-class SdcoreUpfOperatorCharm(ops.CharmBase):
+class SdcoreUpfCharm(ops.CharmBase):
     """Machine charm for SD-Core User Plane Function."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-
-    def _on_config_changed(self, event):
-        self.unit.status = ops.ActiveStatus()
+        self.framework.observe(self.on.install, self._configure)
+        self.framework.observe(self.on.update_status, self._configure)
+        self.framework.observe(self.on.config_changed, self._configure)
+
+    def _configure(self, _):
+        """Handle UPF installation."""
+        if not self.unit.is_leader():
+            self.unit.status = BlockedStatus("Scaling is not implemented for this charm")
+            return
+        self.unit.status = MaintenanceStatus("Installing the UPF")
+        self._install_upf_snap()
+        self.unit.status = ActiveStatus()
```

Comment:
This will go to `MaintenanceStatus` and try to install the snap on all `update_status` and `config_changed`; this might be a bit too much.

Reply:
We don't really  reinstall the snap on every event though but we do check if it is installed. What I do like about this approach is that if the user does anything (ex. remove the snap or change its revision), the charm code will bring it back to what we expect in terms of state. What would be your recommendation?

Reply:
You are right that it will ensure the snap is installed and as we specify the revision, it should not cause issue.

The maintenance message however might be a bit misleading in that case, as everytime there is an event, it will go say it is installing the UPF. Does the lib provide a way to just check that the snap is already installed? Otherwise, just changing the message to something like "Ensuring the UPF is installed" would be enough.

Reply:
Agreed, I just got rid of the maintenance status altogether here.

Reply:
As for the lib question, the method we are calling (ensure) does that, it ensures that "a snap is in a given state."

Reply:
Great, that works for me.